packages/core/src/lib/implementation/compare-scorables.ts

```diff
@@ -0,0 +1,194 @@
+  reports: ReportsToCompare,
+): ReportsDiff['groups'] {
+  const { pairs, added, removed } = matchArrayItemsByKey({
+    before: listGroupsFromAllPlugins(reports.before),
+    after: listGroupsFromAllPlugins(reports.after),
+    key: ({ plugin, group }) => `${plugin.slug}/${group.slug}`,
+  });
+  const { changed, unchanged } = comparePairs(
+    pairs,
+    ({ before, after }) => before.group.score === after.group.score,
+  );
+
+  return {
+    changed: changed.map(pluginGroupPairToDiff),
+    unchanged: unchanged.map(pluginGroupToResult),
+    added: added.map(pluginGroupToResult),
+    removed: removed.map(pluginGroupToResult),
+  };
+}
+
+export function compareAudits(
+  reports: ReportsToCompare,
+): ReportsDiff['audits'] {
+  const { pairs, added, removed } = matchArrayItemsByKey({
+    before: listAuditsFromAllPlugins(reports.before),
+    after: listAuditsFromAllPlugins(reports.after),
+    key: ({ plugin, audit }) => `${plugin.slug}/${audit.slug}`,
+  });
+  const { changed, unchanged } = comparePairs(
+    pairs,
+    ({ before, after }) =>
+      before.audit.value === after.audit.value &&
+      before.audit.score === after.audit.score,
+  );
+
+  return {
+    changed: changed.map(pluginAuditPairToDiff),
+    unchanged: unchanged.map(pluginAuditToResult),
+    added: added.map(pluginAuditToResult),
+    removed: removed.map(pluginAuditToResult),
+  };
+}
+
+function categoryToResult(category: ScoredCategoryConfig): CategoryResult {
+  return {
+    slug: category.slug,
+    title: category.title,
+    score: category.score,
+  };
+}
+
+function categoryPairToDiff({
+  before,
+  after,
+}: Diff<ScoredCategoryConfig>): CategoryDiff {
+  return {
+    slug: after.slug,
+    title: after.title,
+    scores: {
+      before: before.score,
+      after: after.score,
+      diff: after.score - before.score,
+    },
+  };
+}
+
+type PluginGroup = {
+  group: ScoredGroup;
+  plugin: PluginMeta;
+};
+
+function pluginGroupToResult({ group, plugin }: PluginGroup): GroupResult {
+  return {
+    slug: group.slug,
+    title: group.title,
+    plugin: {
+      slug: plugin.slug,
+      title: plugin.title,
+    },
+    score: group.score,
+  };
+}
+
+function pluginGroupPairToDiff({
+  before,
+  after,
+}: Diff<PluginGroup>): GroupDiff {
+  return {
+    slug: after.group.slug,
+    title: after.group.title,
+    plugin: {
+      slug: after.plugin.slug,
+      title: after.plugin.title,
+    },
+    scores: {
+      before: before.group.score,
+      after: after.group.score,
+      diff: after.group.score - before.group.score,
+    },
+  };
+}
+
+type PluginAudit = {
+  audit: AuditReport;
+  plugin: PluginMeta;
+};
+
+function pluginAuditToResult({ audit, plugin }: PluginAudit): AuditResult {
+  return {
+    slug: audit.slug,
+    title: audit.title,
+    plugin: {
+      slug: plugin.slug,
+      title: plugin.title,
+    },
+    score: audit.score,
+    value: audit.value,
+    displayValue: audit.displayValue,
```

Comment:
Shouldn't this be optional? If the `displayValue` is `undefined`, the property shouldn't be there no?

Reply:
It doesn't really matter. It just serialized as JSON in the end, so the end result will be the same - `undefined` isn't supported by JSON format, so `JSON.stringify` will omit such properties.